### ISSUE TITLE
NaN's in original maps

### ISSUE THREAD

**Issue by TomMaullin** [NeuroimagingMetaAnalysis/ibma#4]
Currently the diagnosis program does not account for NaN's in the original contrast estimate and contrast error maps. This leads to NaN's further on and can cause errors such as funnels not being plotted correctly and voxels not having the correct value.

**Comment by AlexBowring** [comment]
For now, the practical way forward is to only compute diagnostic statistics on voxels with no-NaN's (no missing data).  (Note, when you can assume all subjects have data, the test statistic computations can almost always be vectorised -- computable as a single expression for the whole image/slice; this will not be the case once you allow 1 or more missing studies).

However, if it is possible to also compute for voxels that have at least some (e.g. >50% ) of observations it would be nice, but this is secondary.

Finally, an image-wise statistic map that you should compute is the % of studies with data at a given voxel.  If all studies used the exact same analysis map, this would be a boring image, 0% outside the mask, 100% inside.  But most likely this will never happen and there will be a range of '% present'.  ALSO note that FSL provides a useful summary marking the voxels where *just* 1 or *just* 2 studies/subjects are missing... this is useful, because it could be that dropping 1 study could give you a large portion of brack 'back'.

**Comment by nicholst** [comment]
Ach! I'm stilled logged into Github as Alex! Sorry about that!

**Comment by TomMaullin** [comment]
Ah I did wonder! Yes, I was debating creating a 'present study' count map but was worried it would be too much memory for a statistic that might not be used too much. 

If I understand correctly, I had given some thought to what you said about it being harder to do the calculations with missing data. One thought I did have is that in any calculation I could simply temporarily replace NaN's and their corresponding values in other vectors with zeros and the vector calculations would still run correctly so long as I had recorded how many studies had been used (i.e. in the 'present study' map). I believe the statistics that are calculated would still be correct throughout the entire pipeline if I did this carefully.

However, if you don't want the voxels with less than 50% observations perhaps I should give this some further thought - perhaps just multiplying by an indicator somewhere along the line could fix that.

Would it be best to use NaN's or zeros for voxels with no statistic value due to less statistical data than desirable? I believe SPM and FSL have different policies on this?

**Comment by nicholst** [comment]
You are right, that you may be able to come up with a clever way to compute
a statistic in the presence of NaN's.  For example, if you were computing
the mean and you converted NaN's to zeros, you could do
`sum(studies)/sum(studies not NaN)` *however* you need to be careful and
just consider what needs to happen for each type of statistic you're
computing.

Does this make sense?

About the 50%, the concern is that if you only have 1 or 2 studies
contributing the test statistic is going to be junk... so you need some
practical threshold (e.g. it could be expressed in absolute numbers and not
percent... e.g. at least 10 studies) to ensure you have reasonable
statistic values.

**Comment by TomMaullin** [comment]
Yes, I see what you mean. I think this shouldn't be too hard to implement. For now I will stick with greater than the minimum of 50% or 10. This should cover all bases then.

I have worked through the current statistics and think this should be fine to implement. Once the current PR has had an initial review I will give this an attempt.

**Comment by cmaumet** [comment]
Just another quick thing that we will have to take into account: missing values in FSL are stored as zeros instead of NaNs. So, at some point, a test on the analysis software will have to be done in order to interpret correctly where data are missing.

**Comment by nicholst** [comment]
The safest thing is to always use an explicit mask image or, failing that,
some image that you can be sure will never have meaningful zero values
(i.e. variance/standard deviation).
